File: src/grove/claude.py
"""Claude Code memory sync for worktrees.

Claude Code stores per-project memory at ``~/.claude/projects/<encoded-path>/memory/``.
Since worktrees live at different paths than source repos, each gets its own
isolated memory directory. This module syncs memory between source repos and
their worktrees so context isn't lost when worktrees are created or deleted.
"""

from __future__ import annotations

import shutil
from pathlib import Path

CLAUDE_PROJECTS_DIR = Path.home() / ".claude" / "projects"


def encode_path(path: Path) -> str:
    """Encode an absolute path the way Claude Code does.

    Replaces ``/`` and ``.`` with ``-`` so the result is a flat directory name.

    >>> encode_path(Path("/Users/nick/.grove/workspaces/feat-foo"))
    '-Users-nick--grove-workspaces-feat-foo'
    """
    return str(path).replace("/", "-").replace(".", "-")


def memory_dir_for(path: Path) -> Path:
    """Return the Claude Code memory directory for a given project path."""
    return CLAUDE_PROJECTS_DIR / encode_path(path.resolve()) / "memory"


def rehydrate_memory(source_repo: Path, worktree_path: Path) -> int:
    """Copy memory files from *source_repo*'s Claude dir into *worktree_path*'s.

    Creates the target memory directory if it doesn't exist.
    Only copies files that don't already exist in the target (fresh worktree).

    Returns the number of files copied.
    """
    src = memory_dir_for(source_repo)
    if not src.is_dir():
        return 0

    dst = memory_dir_for(worktree_path)
    dst.mkdir(parents=True, exist_ok=True)

    copied = 0
    for src_file in src.iterdir():
        if not src_file.is_file():
            continue
        dst_file = dst / src_file.name
        if not dst_file.exists():
            shutil.copy2(src_file, dst_file)
            copied += 1
    return copied


def harvest_memory(worktree_path: Path, source_repo: Path) -> int:
    """Merge memory files from *worktree_path*'s Claude dir back into *source_repo*'s.

    Copies files that are new or newer (by mtime) than what exists in the
    source repo's memory directory.

    Returns the number of files copied/updated.
    """
    src = memory_dir_for(worktree_path)
    if not src.is_dir():
        return 0

    dst = memory_dir_for(source_repo)
    dst.mkdir(parents=True, exist_ok=True)

    copied = 0
    for src_file in src.iterdir():
        if not src_file.is_file():
            continue
        dst_file = dst / src_file.name
        if not dst_file.exists() or src_file.stat().st_mtime > dst_file.stat().st_mtime:
            shutil.copy2(src_file, dst_file)
            copied += 1
    return copied


def find_orphaned_memory_dirs(workspace_dir: Path) -> list[Path]:
    """Find Claude memory dirs that look like Grove worktrees but no longer exist.

    Builds a set of encoded paths for all *existing* workspace subdirectories,
    then flags any Claude project dir under the workspace prefix that isn't in
    that active set.  This avoids the lossy-decode problem entirely.

    Returns a list of orphaned Claude project directories (the parent of ``memory/``).
    """
    if not CLAUDE_PROJECTS_DIR.is_dir():
        return []

    prefix = encode_path(workspace_dir.resolve())

    # Build set of encoded paths for all existing workspace dirs and their children
    active_encoded: set[str] = set()
    if workspace_dir.is_dir():
        for ws_dir in workspace_dir.iterdir():
            if not ws_dir.is_dir():
                continue
            active_encoded.add(encode_path(ws_dir.resolve()))
            for child in ws_dir.iterdir():
                if child.is_dir():
                    active_encoded.add(encode_path(child.resolve()))

    orphaned: list[Path] = []
    for entry in CLAUDE_PROJECTS_DIR.iterdir():
        if not entry.is_dir():
            continue
        if not entry.name.startswith(prefix):
            continue
        # Must be a subdirectory of workspace_dir (not workspace_dir itself)
        if entry.name == prefix:
            continue
        # Must have a memory/ subdir (otherwise it's not a Claude memory project)
        if not (entry / "memory").is_dir():
            continue
        # If not in the active set, it's orphaned
        if entry.name not in active_encoded:
            orphaned.append(entry)

    return orphaned


def cleanup_orphaned_memory_dirs(dirs: list[Path]) -> int:
    """Remove orphaned Claude project directories.

    Returns the number of directories removed.
    """
    removed = 0
    for d in dirs:
        if d.is_dir():
            shutil.rmtree(d, ignore_errors=True)
            if not d.exists():
                removed += 1
    return removed

File: go/internal/claude/claude.go
// Package claude handles Claude Code memory synchronization between
// source repos and worktrees.
//
// Claude Code stores per-project memory at ~/.claude/projects/<encoded-path>/memory/.
// Since worktrees have different filesystem paths, they get separate memory dirs.
// This package bridges them by copying memory files between source repos and worktrees.
package claude

import (
	"io"
	"os"
	"path/filepath"
	"strings"
)

// EncodePath encodes a filesystem path the same way Claude Code does:
// replace "/" and "." with "-".
func EncodePath(path string) string {
	s := strings.ReplaceAll(path, "/", "-")
	s = strings.ReplaceAll(s, ".", "-")
	return s
}

// MemoryDirFor returns the memory directory for a given path.
// claudeDir is typically ~/.claude.
func MemoryDirFor(claudeDir, path string) string {
	encoded := EncodePath(path)
	return filepath.Join(claudeDir, "projects", encoded, "memory")
}

// RehydrateMemory copies memory files from source repo to worktree.
// Only copies files that don't already exist in the worktree.
// Returns the number of files copied.
func RehydrateMemory(claudeDir, sourceRepo, worktreePath string) int {
	srcDir := MemoryDirFor(claudeDir, sourceRepo)
	dstDir := MemoryDirFor(claudeDir, worktreePath)

	entries, err := os.ReadDir(srcDir)
	if err != nil {
		return 0
	}

	count := 0
	for _, entry := range entries {
		if entry.IsDir() {
			continue
		}
		dstPath := filepath.Join(dstDir, entry.Name())
		if _, err := os.Stat(dstPath); err == nil {
			// Already exists — skip
			continue
		}

		srcPath := filepath.Join(srcDir, entry.Name())
		if copyFile(srcPath, dstPath) == nil {
			count++
		}
	}
	return count
}

// HarvestMemory copies new/newer memory files from worktree back to source.
// Uses mtime comparison: only copies if worktree file is newer than source copy.
// Returns the number of files copied.
func HarvestMemory(claudeDir, worktreePath, sourceRepo string) int {
	wtDir := MemoryDirFor(claudeDir, worktreePath)
	srcDir := MemoryDirFor(claudeDir, sourceRepo)

	entries, err := os.ReadDir(wtDir)
	if err != nil {
		return 0
	}

	count := 0
	for _, entry := range entries {
		if entry.IsDir() {
			continue
		}
		wtPath := filepath.Join(wtDir, entry.Name())
		srcPath := filepath.Join(srcDir, entry.Name())

		wtInfo, err := os.Stat(wtPath)
		if err != nil {
			continue
		}

		// Check if source exists and is newer
		srcInfo, err := os.Stat(srcPath)
		if err == nil && srcInfo.ModTime().After(wtInfo.ModTime()) {
			// Source is newer — don't overwrite
			continue
		}
		if err == nil && srcInfo.ModTime().Equal(wtInfo.ModTime()) {
			continue
		}

		if copyFile(wtPath, srcPath) == nil {
			count++
		}
	}
	return count
}

// MigrateMemoryDir moves a memory directory from oldPath to newPath.
// If the target already exists, merges files from old into new.
// Returns true if migration succeeded or source didn't exist.
func MigrateMemoryDir(claudeDir, oldPath, newPath string) bool {
	oldProjectDir := filepath.Join(claudeDir, "projects", EncodePath(oldPath))
	newProjectDir := filepath.Join(claudeDir, "projects", EncodePath(newPath))

	if _, err := os.Stat(oldProjectDir); os.IsNotExist(err) {
		return false
	}

	// Check if target exists
	if _, err := os.Stat(newProjectDir); err == nil {
		// Merge: copy files from old memory/ to new memory/
		oldMemDir := filepath.Join(oldProjectDir, "memory")
		newMemDir := filepath.Join(newProjectDir, "memory")
		os.MkdirAll(newMemDir, 0o755)

		entries, _ := os.ReadDir(oldMemDir)
		for _, entry := range entries {
			if entry.IsDir() {
				continue
			}
			dst := filepath.Join(newMemDir, entry.Name())
			if _, err := os.Stat(dst); err == nil {
				continue // don't overwrite existing
			}
			copyFile(filepath.Join(oldMemDir, entry.Name()), dst)
		}
		os.RemoveAll(oldProjectDir)
	} else {
		// Simple rename
		os.MkdirAll(filepath.Dir(newProjectDir), 0o755)
		if err := os.Rename(oldProjectDir, newProjectDir); err != nil {
			return false
		}
	}

	return true
}

// FindOrphanedMemoryDirs finds Claude project dirs for worktree paths that
// no longer exist. Uses the workspace dir prefix to scope the search (avoids
// flagging non-grove Claude projects), and cross-references against active
// worktree paths rather than attempting lossy path reconstruction.
func FindOrphanedMemoryDirs(claudeDir string, workspaceDir string) []string {
	projectsDir := filepath.Join(claudeDir, "projects")
	entries, err := os.ReadDir(projectsDir)
	if err != nil {
		return nil
	}

	// Encoded workspace dir prefix for scoping
	wsDirEncoded := EncodePath(workspaceDir)

	// Walk the workspace dir to find all paths that actually exist
	activePaths := make(map[string]bool)
	filepath.Walk(workspaceDir, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return nil
		}
		if info.IsDir() {
			activePaths[EncodePath(path)] = true
		}
		return nil
	})

	var orphaned []string
	for _, entry := range entries {
		if !entry.IsDir() {
			continue
		}
		name := entry.Name()
		projDir := filepath.Join(projectsDir, name)

		// Only consider dirs that have a memory/ subdirectory
		memDir := filepath.Join(projDir, "memory")
		if _, err := os.Stat(memDir); os.IsNotExist(err) {
			continue
		}

		// Only consider dirs scoped to the workspace directory
		if !strings.HasPrefix(name, wsDirEncoded) {
			continue
		}

		// If this encoded path matches an active directory, it's not orphaned
		if activePaths[name] {
			continue
		}

		orphaned = append(orphaned, projDir)
	}

	return orphaned
}

// copyFile copies src to dst atomically, creating parent directories as needed.
func copyFile(src, dst string) error {
	os.MkdirAll(filepath.Dir(dst), 0o755)

	in, err := os.Open(src)
	if err != nil {
		return err
	}
	defer in.Close()

	tmp := dst + ".tmp"
	out, err := os.Create(tmp)
	if err != nil {
		return err
	}

	if _, err := io.Copy(out, in); err != nil {
		out.Close()
		os.Remove(tmp)
		return err
	}
	out.Close()

	return os.Rename(tmp, dst)
}
